src/solvers/refinement/string_refinement.cpp

```diff
@@ -303,70 +323,44 @@ void string_refinementt::concretize_string(const exprt &expr)
 {
   if(refined_string_typet::is_refined_string_type(expr.type()))
   {
-    string_exprt str=to_string_expr(expr);
-    exprt length=get(str.length());
+    const string_exprt str=to_string_expr(expr);
+    const exprt length=get(str.length());
     exprt content=str.content();
     replace_expr(symbol_resolve, content);
     found_length[content]=length;
-    mp_integer found_length;
-    if(!to_integer(length, found_length))
-    {
-      INVARIANT(
-        found_length.is_long(),
-        string_refinement_invariantt("the length of a string should be a "
-          "long"));
-      INVARIANT(
-        found_length>=0,
-        string_refinement_invariantt("the length of a string should be >= 0"));
-      size_t concretize_limit=found_length.to_long();
-      INVARIANT(
-        concretize_limit<=generator.max_string_length,
-        string_refinement_invariantt("string length must be less than the max "
-          "length"));
-      exprt content_expr=str.content();
-      std::vector<exprt> result;
-
-      if(index_set[str.content()].empty())
-        return;
+    const auto string_length=expr_cast<std::size_t>(length);
```

Comment:
Even if this is a genuine check (that cannot be avoided by fixing an alleged bug in the solver -- see below) I'd rather refactor this function to `check_index` (or similar) and make it return a boolean so that exceptions are not used as part of normal operation. Exceptions should be used for exceptional behaviour only.